{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "\n",
    "from PhicoreModules import parse_genbank, median, mean, stdev, mode\n",
    "from PhicoreModules import get_features_of_type, get_features_lengths, get_gc_content, get_coding_density, get_distribution_of_stops # Przemek's functions\n",
    "from PhicoreModules import get_mean_cds_length_rec_window, get_rolling_gc, get_rolling_mean_cds # George's functions\n",
    "from PhicoreModules import write_df_to_artemis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Work in progress on STOP codons entropy per frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Union\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def get_codon_num_in_frame(seq: Union[str, SeqRecord, Seq], codon: str) -> int:\n",
    "    \"\"\"\n",
    "    Count the number of codons in a sequence in the first frame.\n",
    "    :param: seq: nucleotide sequence\n",
    "    :param: codon: codon to count\n",
    "    :return: number of codons\n",
    "    \"\"\"\n",
    "\n",
    "    codons_num = 0\n",
    "    for i in range(0, len(seq), 3):\n",
    "        if seq[i:i+3] == codon:\n",
    "            codons_num += 1\n",
    "    return codons_num\n",
    "\n",
    "\n",
    "def get_distribution_of_stops_per_frame(seqiorec, window=210, step=30):\n",
    "    \"\"\"\n",
    "    Get distribution of stops\n",
    "    :param seqiorec:\n",
    "    :param window:\n",
    "    :param step:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    stops = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "    stops_frame_distr = {\n",
    "        'x': range(1, len(seqiorec.seq) + 1),\n",
    "\n",
    "        'frame1-TAA': [np.NAN]*int(window/2),\n",
    "        'frame1-TAG': [np.NAN]*int(window/2),\n",
    "        'frame1-TGA': [np.NAN]*int(window/2),\n",
    "\n",
    "        'frame2-TAA': [np.NAN]*int(window/2),\n",
    "        'frame2-TAG': [np.NAN]*int(window/2),\n",
    "        'frame2-TGA': [np.NAN]*int(window/2),\n",
    "\n",
    "        'frame3-TAA': [np.NAN]*int(window/2),\n",
    "        'frame3-TAG': [np.NAN]*int(window/2),\n",
    "        'frame3-TGA': [np.NAN]*int(window/2)\n",
    "    }\n",
    "    \n",
    "    i = 0\n",
    "    while i + window/2 + 3 <= len(seqiorec.seq) - window/2:\n",
    "        window_seq = seqiorec.seq[i : i + window]\n",
    "        for frame in range(3):\n",
    "            taa = get_codon_num_in_frame(seqiorec.seq[i + frame: i + window + frame], 'TAA')\n",
    "            tag = get_codon_num_in_frame(seqiorec.seq[i + frame : i + window + frame],'TAG')\n",
    "            tga = get_codon_num_in_frame(seqiorec.seq[i + frame : i + window + frame],'TGA')\n",
    "            stops_frame_distr['frame{}-TAA'.format(frame + 1)].extend([taa]*(step))\n",
    "            stops_frame_distr['frame{}-TAG'.format(frame + 1)].extend([tag]*(step))\n",
    "            stops_frame_distr['frame{}-TGA'.format(frame + 1)].extend([tga]*(step))\n",
    "        i += step\n",
    "        \n",
    "    i -= step\n",
    "    left = len(seqiorec.seq) - len(stops_frame_distr['frame1-TAA'])\n",
    "    if left > 0:\n",
    "        for frame in range(3):\n",
    "            stops_frame_distr['frame{}-TAA'.format(frame + 1)].extend([np.NAN]*left)\n",
    "            stops_frame_distr['frame{}-TAG'.format(frame + 1)].extend([np.NAN]*left)\n",
    "            stops_frame_distr['frame{}-TGA'.format(frame + 1)].extend([np.NAN]*left)\n",
    "\n",
    "    return pd.DataFrame(stops_frame_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "def get_all_stats(infile : str, outdir : str, plot : List[str] = 0,  write : List[str] = []):\n",
    "    \"\"\"\n",
    "    Prints all stats we have so far on a single file\n",
    "    :param infile: path to input file\n",
    "    :param outdir: path to output directory\n",
    "    :param plot: which plots to generate\n",
    "    :param write: which stats and plots to write to file\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    stats = {\n",
    "        'infile': infile,\n",
    "        'id': '', \n",
    "        'length': 0, \n",
    "        'gc': 0, \n",
    "        'cds_num': 0, \n",
    "        'cds_avg_len': 0, \n",
    "        'cds_med_len': 0, \n",
    "        'cds_tRNA_density': 0, \n",
    "\n",
    "    }\n",
    "\n",
    "    print(\"=== Working on file {} ===\".format(infile))\n",
    "    for record in parse_genbank(infile):\n",
    "        stats['id'] = record.id\n",
    "        stats['length'] = len(record.seq)\n",
    "        stats['gc'] = get_gc_content(record.seq)\n",
    "        cdss = get_features_of_type(record, 'CDS')\n",
    "        cdss_lengths = get_features_lengths(record, 'CDS')\n",
    "        stats['cds_num'] = len(cdss)\n",
    "        stats['cds_avg_len'] = mean(cdss_lengths)\n",
    "        stats['cds_med_len'] = median(cdss_lengths)\n",
    "        stats['cds_tRNA_density'] = get_coding_density(record, ['CDS', 'tRNA'])\n",
    "\n",
    "        # output files\n",
    "        df_stops_file = os.path.join(outdir, '{}.{}_stops.txt'.format(os.path.basename(infile), record.id))\n",
    "        df_stops_per_frame_file = os.path.join(outdir, '{}.{}_stops_per_frame.txt'.format(os.path.basename(infile), record.id))\n",
    "        df_stops_plot_file = os.path.join(outdir, '{}.{}_stops.png'.format(os.path.basename(infile), record.id))\n",
    "        df_stops_per_frame_plot_file = os.path.join(outdir, '{}.{}_stops_per_frame.png'.format(os.path.basename(infile), record.id))\n",
    "        # data\n",
    "        window = 300\n",
    "        step = 30\n",
    "\n",
    "        if plot or write:\n",
    "            print(\" = Getting overall distribution of stops\")\n",
    "            df = get_distribution_of_stops(record, window, step)\n",
    "            if 'stop_distr' in plot:\n",
    "                df.plot(\n",
    "                    x=\"x\", \n",
    "                    y=df.columns[1:], \n",
    "                    figsize=(40, 8), \n",
    "                    legend=True, \n",
    "                    title=\"{} - {} # window: {}, step: {}\".format(os.path.basename(infile), record.id, window, step)\n",
    "                )\n",
    "            if 'stop_distr' in write:\n",
    "                print(\"Writing distribution of stops to {}\".format(df_stops_file))\n",
    "                write_df_to_artemis(df, df_stops_file)\n",
    "                print(\"Writing plot of the distribution to {}\".format(df_stops_plot_file))\n",
    "                plt.savefig(df_stops_plot_file)\n",
    "\n",
    "            if 'stop_distr_per_frame' in plot:\n",
    "                print(\" = Getting distribution of stops per frame\")\n",
    "                df = get_distribution_of_stops_per_frame(record, window, step)\n",
    "                df.plot(\n",
    "                    x=\"x\", \n",
    "                    y=df.columns[1:], \n",
    "                    figsize=(40, 8), \n",
    "                    legend=True, \n",
    "                    title=\"{} - {} # window: {}, step: {}\".format(os.path.basename(infile), record.id, window, step)\n",
    "                )\n",
    "            if 'stop_distr_per_frame' in write:\n",
    "                print(\"Writing distribution of stops to {}\".format(df_stops_file))\n",
    "                write_df_to_artemis(df, df_stops_per_frame_file)\n",
    "                print(\"Writing plot of the distribution to {}\".format(df_stops_plot_file))\n",
    "                plt.savefig(df_stops_per_frame_plot_file)\n",
    "    \n",
    "    return stats\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Working on file ./genbank/Bc11.fasta-TAA.gbk ===\n",
      "=== Working on file ./genbank/Bc11.fasta-TAG-TAA.gbk ===\n",
      "=== Working on file ./genbank/Bc11.fasta-TAG-TGA-TAA.gbk ===\n",
      "=== Working on file ./genbank/Bc11.fasta-TAG-TGA.gbk ===\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Working on file ./genbank/Bc11.fasta-TAG.gbk ===\n",
      "=== Working on file ./genbank/Bc11.fasta-TGA-TAA.gbk ===\n",
      "=== Working on file ./genbank/Bc11.fasta-TGA.gbk ===\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       contig_1_rc     90579 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_9367/2247737863.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>infile</th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>gc</th>\n",
       "      <th>cds_num</th>\n",
       "      <th>cds_avg_len</th>\n",
       "      <th>cds_med_len</th>\n",
       "      <th>cds_tRNA_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>./genbank/Bc11.fasta-TAA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>102</td>\n",
       "      <td>287.245098</td>\n",
       "      <td>114.0</td>\n",
       "      <td>0.960752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>./genbank/Bc11.fasta-TAG-TAA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>107</td>\n",
       "      <td>273.140187</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.958511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>./genbank/Bc11.fasta-TAG-TGA-TAA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>101</td>\n",
       "      <td>286.801980</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.954305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>./genbank/Bc11.fasta-TAG-TGA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>99</td>\n",
       "      <td>300.161616</td>\n",
       "      <td>121.0</td>\n",
       "      <td>0.969850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>./genbank/Bc11.fasta-TAG.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>82</td>\n",
       "      <td>368.036585</td>\n",
       "      <td>169.0</td>\n",
       "      <td>0.974575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>./genbank/Bc11.fasta-TGA-TAA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>106</td>\n",
       "      <td>278.254717</td>\n",
       "      <td>118.0</td>\n",
       "      <td>0.965721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>./genbank/Bc11.fasta-TGA.gbk</td>\n",
       "      <td>contig_1_rc</td>\n",
       "      <td>90579</td>\n",
       "      <td>0.291458</td>\n",
       "      <td>80</td>\n",
       "      <td>376.012500</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.979995</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 infile           id  length        gc  \\\n",
       "0          ./genbank/Bc11.fasta-TAA.gbk  contig_1_rc   90579  0.291458   \n",
       "1      ./genbank/Bc11.fasta-TAG-TAA.gbk  contig_1_rc   90579  0.291458   \n",
       "2  ./genbank/Bc11.fasta-TAG-TGA-TAA.gbk  contig_1_rc   90579  0.291458   \n",
       "3      ./genbank/Bc11.fasta-TAG-TGA.gbk  contig_1_rc   90579  0.291458   \n",
       "4          ./genbank/Bc11.fasta-TAG.gbk  contig_1_rc   90579  0.291458   \n",
       "5      ./genbank/Bc11.fasta-TGA-TAA.gbk  contig_1_rc   90579  0.291458   \n",
       "6          ./genbank/Bc11.fasta-TGA.gbk  contig_1_rc   90579  0.291458   \n",
       "\n",
       "   cds_num  cds_avg_len  cds_med_len  cds_tRNA_density  \n",
       "0      102   287.245098        114.0          0.960752  \n",
       "1      107   273.140187        112.0          0.958511  \n",
       "2      101   286.801980        120.0          0.954305  \n",
       "3       99   300.161616        121.0          0.969850  \n",
       "4       82   368.036585        169.0          0.974575  \n",
       "5      106   278.254717        118.0          0.965721  \n",
       "6       80   376.012500        211.0          0.979995  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "infiles = glob.glob('./genbank/Bc11*')\n",
    "outdir = './tmp'\n",
    "df = pd.DataFrame()\n",
    "for infile in infiles:\n",
    "    df = df.append(get_all_stats(infile, outdir), ignore_index=True)\n",
    "\n",
    "df\n",
    "\n",
    "# for record in parse_genbank('./genbank/OFRY01000050-TAG-TGA-TAA.gbk'):\n",
    "#     df_all = get_rolling_mean_cds(record, window=1000, step=30)\n",
    "# for record in parse_genbank('./genbank/OFRY01000050-TGA-TAA.gbk'):\n",
    "#     df_tag = get_rolling_mean_cds(record, window=1000, step=30)\n",
    "# for record in parse_genbank('./genbank/OFRY01000050-TAG-TAA.gbk'):\n",
    "#     df_tga = get_rolling_mean_cds(record, window=1000, step=30)\n",
    "# for record in parse_genbank('./genbank/OFRY01000050-TAG-TGA.gbk'):\n",
    "#     df_taa = get_rolling_mean_cds(record, window=1000, step=30)\n",
    "\n",
    "\n",
    "# df_all['No_TAG'] = df_tag['Mean_CDS']\n",
    "# df_all['No_TGA'] = df_tga['Mean_CDS']\n",
    "# df_all['No_TAA'] = df_taa['Mean_CDS']\n",
    "\n",
    "# # df_tag['Mean_CDS_all'] = df_all['Mean_CDS']\n",
    "# df_all.plot(x=\"x\", y=[\"Mean_CDS\", \"No_TAG\", \"No_TGA\", \"No_TAA\"], figsize=(40, 8))\n",
    "\n",
    "# # df_tag['Delta'] = df_tag['Mean_CDS'] - df_tag['Mean_CDS_all']\n",
    "\n",
    "# outfile = './tmp/OFRY01000050-TAG-TGA-TAA.gbk.cds_distr_w1000_s30.txt'\n",
    "# write_df_to_artemis(df_all, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get some stats for a single genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x 96512 range(1, 11) range(96503, 96513)\n",
      "TAA 96532 [nan, nan, nan, nan, nan, nan, nan, 14, 14, 14] [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "TAG 96532 [nan, nan, nan, nan, nan, nan, nan, 5, 5, 5] [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "TGA 96532 [nan, nan, nan, nan, nan, nan, nan, 2, 2, 2] [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dec/miniconda3/lib/python3.9/site-packages/Bio/GenBank/Scanner.py:1555: BiopythonParserWarning: Malformed LOCUS line found - is this correct?\n",
      ":'LOCUS       OFRY01000050     96512 bp    DNA             PHG\\n'\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "All arrays must be of the same length",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb Komórka 7\u001b[0m in \u001b[0;36m<cell line: 46>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=45'>46</a>\u001b[0m \u001b[39mfor\u001b[39;00m infile \u001b[39min\u001b[39;00m infiles[:\u001b[39m1\u001b[39m]:\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=46'>47</a>\u001b[0m     \u001b[39mfor\u001b[39;00m record \u001b[39min\u001b[39;00m parse_genbank(infile):\n\u001b[0;32m---> <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=47'>48</a>\u001b[0m         df \u001b[39m=\u001b[39m get_distribution_of_stops(record,\u001b[39m300\u001b[39;49m, \u001b[39m15\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=48'>49</a>\u001b[0m         \u001b[39mprint\u001b[39m(df\u001b[39m.\u001b[39mtail())\n",
      "\u001b[1;32m/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb Komórka 7\u001b[0m in \u001b[0;36mget_distribution_of_stops\u001b[0;34m(seqiorec, window, step)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=40'>41</a>\u001b[0m \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m stops_distr\u001b[39m.\u001b[39mitems():\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=41'>42</a>\u001b[0m     \u001b[39mprint\u001b[39m(k, \u001b[39mlen\u001b[39m(v), v[:\u001b[39m10\u001b[39m], v[\u001b[39m-\u001b[39m\u001b[39m10\u001b[39m:])\n\u001b[0;32m---> <a href='vscode-notebook-cell://wsl%2Bubuntu-20.04/d/analyses/BEKKER/CODON_REASSIGNMENT/Phicore/getting_started.ipynb#X11sdnNjb2RlLXJlbW90ZQ%3D%3D?line=43'>44</a>\u001b[0m \u001b[39mreturn\u001b[39;00m pd\u001b[39m.\u001b[39;49mDataFrame(stops_distr)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/frame.py:636\u001b[0m, in \u001b[0;36mDataFrame.__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    630\u001b[0m     mgr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_init_mgr(\n\u001b[1;32m    631\u001b[0m         data, axes\u001b[39m=\u001b[39m{\u001b[39m\"\u001b[39m\u001b[39mindex\u001b[39m\u001b[39m\"\u001b[39m: index, \u001b[39m\"\u001b[39m\u001b[39mcolumns\u001b[39m\u001b[39m\"\u001b[39m: columns}, dtype\u001b[39m=\u001b[39mdtype, copy\u001b[39m=\u001b[39mcopy\n\u001b[1;32m    632\u001b[0m     )\n\u001b[1;32m    634\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(data, \u001b[39mdict\u001b[39m):\n\u001b[1;32m    635\u001b[0m     \u001b[39m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[39;00m\n\u001b[0;32m--> 636\u001b[0m     mgr \u001b[39m=\u001b[39m dict_to_mgr(data, index, columns, dtype\u001b[39m=\u001b[39;49mdtype, copy\u001b[39m=\u001b[39;49mcopy, typ\u001b[39m=\u001b[39;49mmanager)\n\u001b[1;32m    637\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(data, ma\u001b[39m.\u001b[39mMaskedArray):\n\u001b[1;32m    638\u001b[0m     \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mma\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mmrecords\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mmrecords\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/internals/construction.py:502\u001b[0m, in \u001b[0;36mdict_to_mgr\u001b[0;34m(data, index, columns, dtype, typ, copy)\u001b[0m\n\u001b[1;32m    494\u001b[0m     arrays \u001b[39m=\u001b[39m [\n\u001b[1;32m    495\u001b[0m         x\n\u001b[1;32m    496\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mhasattr\u001b[39m(x, \u001b[39m\"\u001b[39m\u001b[39mdtype\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(x\u001b[39m.\u001b[39mdtype, ExtensionDtype)\n\u001b[1;32m    497\u001b[0m         \u001b[39melse\u001b[39;00m x\u001b[39m.\u001b[39mcopy()\n\u001b[1;32m    498\u001b[0m         \u001b[39mfor\u001b[39;00m x \u001b[39min\u001b[39;00m arrays\n\u001b[1;32m    499\u001b[0m     ]\n\u001b[1;32m    500\u001b[0m     \u001b[39m# TODO: can we get rid of the dt64tz special case above?\u001b[39;00m\n\u001b[0;32m--> 502\u001b[0m \u001b[39mreturn\u001b[39;00m arrays_to_mgr(arrays, columns, index, dtype\u001b[39m=\u001b[39;49mdtype, typ\u001b[39m=\u001b[39;49mtyp, consolidate\u001b[39m=\u001b[39;49mcopy)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/internals/construction.py:120\u001b[0m, in \u001b[0;36marrays_to_mgr\u001b[0;34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[39mif\u001b[39;00m verify_integrity:\n\u001b[1;32m    118\u001b[0m     \u001b[39m# figure out the index, if necessary\u001b[39;00m\n\u001b[1;32m    119\u001b[0m     \u001b[39mif\u001b[39;00m index \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 120\u001b[0m         index \u001b[39m=\u001b[39m _extract_index(arrays)\n\u001b[1;32m    121\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    122\u001b[0m         index \u001b[39m=\u001b[39m ensure_index(index)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/pandas/core/internals/construction.py:674\u001b[0m, in \u001b[0;36m_extract_index\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    672\u001b[0m lengths \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39m(\u001b[39mset\u001b[39m(raw_lengths))\n\u001b[1;32m    673\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(lengths) \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[0;32m--> 674\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mAll arrays must be of the same length\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    676\u001b[0m \u001b[39mif\u001b[39;00m have_dicts:\n\u001b[1;32m    677\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    678\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mMixing dicts with non-Series may lead to ambiguous ordering.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    679\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: All arrays must be of the same length"
     ]
    }
   ],
   "source": [
    "infiles = glob.glob('./genbank/OFRY01000050*')\n",
    "outdir = './tmp'\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "def get_distribution_of_stops(seqiorec: SeqRecord, window: int = 210, step: int = 1) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Get distribution of STOP codons in a sequence\n",
    "    :param seqiorec: SeqRecord object\n",
    "    :param window: window size\n",
    "    :param step: step size\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    stops = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "    stops_distr = {\n",
    "        'x': range(1, len(seqiorec.seq) + 1),\n",
    "        'TAA': [np.NAN]*int(step/2),\n",
    "        'TAG': [np.NAN]*int(step/2),\n",
    "        'TGA': [np.NAN]*int(step/2)\n",
    "    }\n",
    "    \n",
    "    i = 0\n",
    "    while i < len(seqiorec.seq):\n",
    "        window_seq = seqiorec.seq[i : i + window]\n",
    "        taa = window_seq.count('TAA')\n",
    "        tag = window_seq.count('TAG')\n",
    "        tga = window_seq.count('TGA')\n",
    "        stops_distr['TAA'].extend([taa]*(step))\n",
    "        stops_distr['TAG'].extend([tag]*(step))\n",
    "        stops_distr['TGA'].extend([tga]*(step))\n",
    "        i += step\n",
    "        \n",
    "    i -= step\n",
    "    left = len(seqiorec.seq) - len(stops_distr['TAA'])\n",
    "    if left > 0:   \n",
    "        print('aa')\n",
    "        stops_distr['TAA'].extend([np.NAN]*left)\n",
    "        stops_distr['TAG'].extend([np.NAN]*left)\n",
    "        stops_distr['TGA'].extend([np.NAN]*left)\n",
    "\n",
    "    for k, v in stops_distr.items():\n",
    "        print(k, len(v), v[:10], v[-10:])\n",
    "\n",
    "    return pd.DataFrame(stops_distr)\n",
    "\n",
    "for infile in infiles[:1]:\n",
    "    for record in parse_genbank(infile):\n",
    "        df = get_distribution_of_stops(record,300, 15)\n",
    "        print(df.tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "08829089b8420cee51176c00dfbb80c517ddb771c0f915ce4701dca739baf8ec"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
